keanu-python/keanu/vartypes.py

```diff
@@ -1,16 +1,65 @@
+from typing import Iterable, Union, Type, Any, TYPE_CHECKING
 import numpy as np
-import pandas as pd
+from pandas import Series, DataFrame
+from keanu.vertex.ops import VertexOps
+from .base import JavaObjectWrapper
 
-int_types = (int, np.integer)
 
-float_types = (float, np.float)
+# see numpy's scalar hierarchy: https://docs.scipy.org/doc/numpy/reference/arrays.scalars.html#scalars
+int_types = Union[int, np.integer]
+float_types = Union[float, np.floating]
+bool_types = Union[bool, np.bool_]
 
-bool_types = (bool, np.bool_)
+primitive_types = Union[int_types, float_types, bool_types]
 
-primitive_types = int_types + float_types + bool_types
+'''
+Tensor arg types
+'''
+pandas_types = Union[Series, DataFrame]
+numpy_types = Union[np.ndarray]
+pandas_and_numpy_types = Union[pandas_types, numpy_types]
 
-pandas_types = (pd.Series, pd.DataFrame)
+# mypy treats bool as a subtype of int and int as a subtype of float
+# see : https://github.com/python/mypy/issues/1850
+tensor_arg_types = Union[primitive_types, pandas_and_numpy_types]
+bool_tensor_arg_types = Union[bool_types, pandas_and_numpy_types]
+int_and_bool_tensor_arg_types = Union[int_types, bool_types, pandas_and_numpy_types]
 
-numpy_types = (np.ndarray, )
+'''
+Vertex arg types
+'''
+wrapped_java_types = Union[JavaObjectWrapper]
+shape_types = Iterable[primitive_types]
 
-const_arg_types = numpy_types + pandas_types + primitive_types
+# mypy treats bool as a subtype of int and int as a subtype of float
+# see : https://github.com/python/mypy/issues/1850
+vertex_param_types = Union[tensor_arg_types, wrapped_java_types, VertexOps]
+bool_vertex_param_types = Union[bool_tensor_arg_types, wrapped_java_types, VertexOps]
+int_and_bool_vertex_param_types = Union[int_and_bool_tensor_arg_types, wrapped_java_types, VertexOps]
+
+vertex_arg_types = Union[vertex_param_types, shape_types]
+
+'''
+Runtime types
+'''
+# mypy prohibits use of static types from typing module at runtime
+# see : https://github.com/python/mypy/issues/5354
+runtime_numpy_types : Type[Any]
+runtime_pandas_types : Type[Any]
+runtime_primitive_types : Type[Any]
+runtime_int_types : Type[Any]
+runtime_float_types : Type[Any]
+runtime_bool_types : Type[Any]
+runtime_tensor_arg_types : Type[Any]
+runtime_wrapped_java_types : Type[Any]
+
+if not TYPE_CHECKING:
+    # Union with one argument does not have __args__ parameter
+    runtime_numpy_types = numpy_types
```

Comment:
I wonder if it would be cleaner to implement this as a method `runtime_safe(numpy_types)` etc

Reply:
I think it's enough to just have the variable name, since it's all under the `if not TYPE_CHECKING` condition

Reply:
I was imagining that a general `runtime_safe(types)` method would mean you don't have to implement a `runtime_` version of lots of different type collections, and it's clear in the calling code what the difference is between `foo` and `runtime_safe(foo)`.

i.e. you could replace lines 36-54 with:

    def runtime_safe(type_collection):
        if TYPE_CHECKING:
            return Type[Any]
        else:
            return type_collection # or I guess return type_collection.__args__ if it exists